Add typing to amazon provider EMR

Swalloow opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

related: #9708 
Adding type annotations to `amazon/emr`.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

Swalloow committed:
Add typing to amazon provider EMR

Swalloow committed:
Add typing to amazon provider EMR

jayrumi committed:
Add get_blobs_list method to WasbHook (#9950)

Swalloow committed:
Fix static check error

Swalloow committed:
Fix static check error

mik-laj commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -65,7 +71,7 @@ def get_cluster_id_by_name(self, emr_cluster_name, cluster_states):
             return None
 
-    def create_job_flow(self, job_flow_overrides):
+    def create_job_flow(self, job_flow_overrides) -> Dict[str, Any]:
```
```suggestion
    def create_job_flow(self, job_flow_overrides: Dict[str, Any]) -> Dict[str, Any]:
```

mik-laj commented on airflow/providers/amazon/aws/operators/emr_add_steps.py:
```diff
@@ -69,13 +70,14 @@ def __init__(
         self.steps = steps
 
-    def execute(self, context):
+    def execute(self, context) -> List[str]:
```
```suggestion
    def execute(self, context: Dict[str, Any]) -> List[str]:
```

mik-laj commented on airflow/providers/amazon/aws/operators/emr_create_job_flow.py:
```diff
@@ -57,7 +58,7 @@ def __init__(
         self.region_name = region_name
 
-    def execute(self, context):
+    def execute(self, context) -> str:
```
```suggestion
    def execute(self, context: Dict[str, Any]) -> str:
```

mik-laj commented on airflow/providers/amazon/aws/operators/emr_modify_cluster.py:
```diff
@@ -51,7 +51,7 @@ def __init__(
         self.step_concurrency_level = step_concurrency_level
 
-    def execute(self, context):
+    def execute(self, context) -> int:
```
```suggestion
    def execute(self, context: Dict[str, Any]) -> int:
```

mik-laj commented on airflow/providers/amazon/aws/operators/emr_terminate_job_flow.py:
```diff
@@ -37,14 +37,14 @@ class EmrTerminateJobFlowOperator(BaseOperator):
         self.aws_conn_id = aws_conn_id
 
-    def execute(self, context):
+    def execute(self, context) -> None:
```
```suggestion
    def execute(self, context: Dict[str, Any]) -> None:
```

kaxil committed:
Fix OpenShift Guidelines link in IMAGES.rst (#9978)

kaxil committed:
Add Badges for Airflow Docker (#9979)

Swalloow committed:
Add missing types

Swalloow committed:
Add missing types

kaxil committed:
Fix typo in airflow/cli/cli_parser.py (#9980)

`action_subcommnads` -> `action_subcommands`
`group_subcommnands` -> `group_subcommands`

mik-laj commented on airflow/providers/amazon/aws/operators/emr_create_job_flow.py:
```diff
@@ -70,7 +71,7 @@ def execute(self, context):
             self.job_flow_overrides = ast.literal_eval(self.job_flow_overrides)
 
-        response = emr.create_job_flow(self.job_flow_overrides)
+        response = emr.create_job_flow(self.job_flow_overrides)     # type: ignore
```
You need to create a new variable. 
```
            job_flow_overrides = ast.literal_eval(self.job_flow_overrides)
        else:
            job_flow_overrides = self.job_flow_overrides
        response = emr.create_job_flow(job_flow_overrides)
```

Swalloow replied:
@mik-laj Thanks for the review. I just fixed it :D

Swalloow committed:
Remove ignore comments

Swalloow committed:
Remove ignore comments

kaxil committed:
Fix various typos in airflow/cli/commands (#9983)

Swalloow committed:
Remove ignore comments

Swalloow committed:
Remove ignore comments

kaxil committed:
Stop using start_date in default_args in example_dags (#9982)

Swalloow committed:
Fix testcase error

Swalloow committed:
Fix testcase error

kaxil committed:
Bump actions/setup-python version to 2 (#9984)

v2 was released on 21 March

kaxil committed:
Bump tableauserverclient to 0.12 (#9988)

kaxil committed:
Relax requirement to allow latest version of flask-caching (#9989)

Current available version is 1.9.0

kaxil committed:
Relax requirement to allow latest version of jinja2 (#9991)

The latest available version of jinja2 is 2.11.2

Fixes in 2.11.2: https://github.com/pallets/jinja/blob/master/CHANGES.rst#version-2112

kaxil committed:
Relax requirement to allow latest version of tenacity (#9992)

kaxil committed:
Bump mysqlclient to 2.0.1 (#9987)

mysqlclient was pinned as the functionality of Connection object of MySQL had changed and that it dropped python2.

mysqlclient has added context manager interface to Connection which closes the connection on `__exit__` in 2.0.0 and Airflow Master no longer uses Python 2

potiuk committed:
Revert "Bump mysqlclient to 2.0.1 (#9987)" (#9997)

This reverts commit c4388127f1c47e1abe512e1ef104c9835a150a04.

Commit:
Remove unnecessary environment variable from CI workflow (#9998)

kaxil committed:
Fix typo in api_connexion/openapi/v1.yaml (#9986)

`startd_ate_lte` -> `start_date_lte`

22quinn committed:
Avoid sharing session with RenderedTaskInstanceFields write and delete (#9993)

Sharing session with RTIF leads to idle-in-transaction timeout error when DAG serialization is enabled and task running duration exceeds the idle-in-transaction timeout setting of the database.

potiuk committed:
Adds separate scheduled-only workflow to cancel duplicates (#9999)

Unfortunately cancelling workflows does not work from the workflows
executed by forks because their tokens do not allow doing that
(they are read only). So we have to run a separate cron-triggered
action (in the context of apache/airflow repository to cancel
all duplicate workflows. This action stops all the workflows
running from the same fork/branch except the last one.

natasharw committed:
Fix typos in README.md (#10000)

This commit fixes three typing errors in README.md, consisting of one 
incorrect spelling of the word released and two incorrect version 
notations.

Commit:
Add missing x-openapi-router-controller to DAG Run endpoint (#9945)

alokshenoy committed:
Updates the slack WebClient call to use the instance variable - token (#9995)

Co-authored-by: Alok Shenoy <ashenoy@coursera.org>
Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

kaxil committed:
Stop using start_date in default_args in example_dags (2) (#9985)

potiuk committed:
Fix cron schedule on cancelling workflow (#10002)

rootcss committed:
Add unit tests for GcpBodyFieldValidator in google cloud providers (#10003)

rootcss committed:
Add unit tests for MsSqlHook (#10006)

ipeluffo commented on airflow/providers/amazon/aws/hooks/emr.py:
```diff
@@ -36,7 +38,11 @@ def __init__(self, emr_conn_id=None, *args, **kwargs):
+    def get_cluster_id_by_name(
+        self,
+        emr_cluster_name: str,
+        cluster_states: Optional[List[str]]
```
Probably this should be:
```python
cluster_states: List[str]
```
However, if it's really optional, then the correct use would be (keep in mind this would change the original function's signature):
```python
cluster_states: Optional[List[str]] = None
```

Swalloow replied:
Thanks, I rechecked the boto3 API and you were right.
Also changed the `cluster_states` in `emr_add_steps.py`.

kaxil committed:
Fix Markdown escape in UPDATING.md (#10010)

rootcss committed:
Add unit tests for GcpBodyFieldSanitizer in Google providers (#9996)

darwinyip committed:
Add typing to ImapHook (#9887)

whrocha committed:
Add EBANX company to README.md (#10012)

Swalloow committed:
Fix cluster_states type to list only

Swalloow committed:
Fix cluster_states type to list only

potiuk committed:
Cancel duplicate runs and HEAD runs that failed at specific jobs (#10008)

This change implements canceling of the workflows in case some of
the important (prerequisite) jobs failed - such as static checks
or docs. The scheduled workflow will run periodically and check
if there are no duplicate workflow runs (it will cancel all but
the most recent one). It will also check for the most recent
ones and will cancel them if one of the specified jobs (matched
by regular expression) failed.

The scenario that we want to handle here:

* we want to start tests as soon as possible so that we do not
  have to wait for static checks and doc builds
* on the other hand as soon as one of the "quicker" jobs fail
  we want to fail whole workflow so that the workers can be
  freed up for other runs.

This has to be done as a scheduled run because forked runs
do not have permissions to cancel any of the other runs in
the main repository.

rootcss committed:
Add unit tests for samba provider (#9959)

mik-laj committed:
Add airflow config get-value command (#9932)

turbaszek committed:
Use consistent message in SchedulerJob._process_executor_events (#9929)

turbaszek committed:
Simplify if clauses in ExternalTaskSensor (#9968)


Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

kaxil committed:
Create a short-link for Airflow Slack Invites (#10034)

turbaszek committed:
Introduce BaseExecutor.validate_command to avoid duplication (#10033)

mik-laj committed:
Move e-mail operator to core (#10013)

feluelle committed:
Set pytest version to be < 6.0.0 due to breaking changes (#10043)

The latest pytest version 6.0.0 released yesterday (2020-07-28) 
does not work in conjunction with the version of pylint (2.4.3) we 
are using.

guptamyr committed:
Adding new SageMaker operator for ProcessingJobs (#9594)

OldPanda committed:
Fix docstrings in BigQueryGetDataOperator (#10042)

chipmyersjr committed:
Add typing for jira provider (#10005)

feluelle committed:
Fix PythonVirtualenvOperator not working with Airflow context (#9394)

- automatically add dill requirement if use_dill=True
- add howto docs
- refactor

Co-authored-by: Luis Magana <maganaluis@users.noreply.github.com>

dstandish committed:
UI Graph View: Focus upstream / downstream task dependencies on mouseover (#9303)

* graph view mouseover task should increase stroke width of upstream / downstream
* mouseover should focus dependencies on mouseover in graph view

kaxil committed:
Burst Virtualenv Cache for Kubernetes Testing (#10070)

yauhenishh committed:
Update writing in timezone.rst (#10066)

The questionable symbol changed in "naive"

loozhengyuan committed:
Improve docstring note about GKEStartPodOperator on KubernetesPodOperator (#10049)

Improve recommendation note for GKEStartPodOperator

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

kaxil committed:
Allow `image` in `KubernetesPodOperator` to be templated (#10068)

fixes https://github.com/apache/airflow/issues/10063

turbaszek committed:
Split Display Video 360 example into smaler DAGs (#10077)

fabien-sarcel committed:
Fix typo on reattach property of kubernetespodoperator (#10056)

coopergillan committed:
Fix typo in Athena sensor retries (#10079)

Understanding that it is an attribute name, which could have downstream
consequences, correct the spelling of max_retries and reword some of the
docstring.

feluelle committed:
Pin google-cloud-kms to ..,<2.0.0 due to breaking changes (#10088)

mik-laj committed:
Group UPDATING.md  entries into sections (#10090)

mik-laj committed:
Move the contribution workflow to the beginning of the file (#10092)

mik-laj committed:
Combine entries in logging configuration section (#10094)

chipmyersjr committed:
Remove `args` parameter from provider operator constructors (#10097)

kaxil committed:
Minor fixes in CONTRIBUTING.rst (#10101)

mik-laj committed:
Move Naming Conventions section in CONTRIBUTING.rst (#10103)

mik-laj committed:
Create "major changes" section in UPDATING.md (#10100)

mik-laj committed:
More user-oriented change titles in Python API sections (#10099)

potiuk committed:
Update .asf.yaml (#10110)

rootcss committed:
Add unit tests for mlengine_prediction_summary (#10022)

olchas committed:
Fix hook not passing gcp_conn_id to base class (#10075)

Co-authored-by: Kamil Olszewski <kamil.olszewski@polidea.com>

olchas committed:
Fix sensor not providing arguments for GCSHook (#10074)

Co-authored-by: Kamil Olszewski <kamil.olszewski@polidea.com>

mik-laj committed:
Combine entries in UPDATING.md file (#10102)

ephraimbuddy committed:
Add migration guide for CLI commands (#10078)

ryanyuan committed:
Add missing params to GCP Pub/Sub creation_subscription (#10106)

Add missing params to GCP Pub/Sub creation_subscription hook/operator

potiuk committed:
Fixes flaky kubernetes Pod Operator tests (#10111)

All Pod Operator tests were using the same task id which made it
flaky as sometimes failed tests were still being deleted while
subsequent success tests already started. That lead to random 404
(not found) errors when the failed test was picked up first.

Hasan-J committed:
Make conn_id unique in Connections table (#9067)

kaxil committed:
[AIRFLOW-4541] Replace os.mkdirs usage with pathlib.Path(path).mkdir (#10117)

`makedirs` is used in `airlfow.utils.file.mkdirs`  - it is replaced with pathlib now with python3.5+

potiuk committed:
Get rid of pydruid limitation (#9965)

Pydruid version 0.5.8 failed on python 3.7 but pydruid 0.5.11
fixed it apparently.

chipmyersjr committed:
Add typing annotations to Segment provider (#10120)

turbaszek committed:
Add try clause to DataFusionHook.wait_for_pipeline_state (#10031)

Sometimes it may happen that the pipeline is not visible instantly in
DataFusion so retrieving it will result in 404
closes: #10030

ephraimbuddy committed:
Move stable REST API migration guide to UPDATING.md (#10098)

turbaszek committed:
Add additional Cloud Datastore operators (#10032)

This PR adds more operators for Google Cloud Datastore
service. It also adds missing tests and how-to guides.

potiuk committed:
Moved webserver background to Quarantine (#10114)

mik-laj committed:
Delete irrelevant entries from UPDATING.md (#10093)

anikarni committed:
Add S3KeysUnchangedSensor (#9817)

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>
Co-authored-by: Felix Uellendall <feluelle@users.noreply.github.com>

ephraimbuddy committed:
Add Legacy command displaying new CLI counterparts (#10115)

coopergillan committed:
[GH-9708] Add type coverage to Sendgrid module (#10134)

Declare a "custom" AddressesType that takes a string or iterable of
strings, which covers to, cc, and bcc, then declare this and other data
types for the sendgrid plugin.

potiuk committed:
Status of quarantined tests is stored in Github Issue (#10119)

potiuk committed:
Fixed mistyped quarantine.yaml extension (#10139)

cjsekl committed:
Improve Typing coverage of amazon/aws/athena (#10025)

Co-authored-by: Johan Eklund <jeklund@zynga.com>
Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

potiuk committed:
Replace file.io with artifacts (#10137)

potiuk committed:
Fixes quarantine parsing teething issues (#10145)

* wrong issue id (from tests)
* comment field was copied from status

potiuk committed:
Retry max 3 times if failing to initialize integration (#10146)

turbaszek committed:
Do not use .format() in log messages (#10150)

auvipy committed:
Update celery and kombu versions (#9496)

coopergillan committed:
Use conn_name_attr for SqliteHook connection (#10156)

The DbApiHook allows for a conn_name_attr to be changed in subclasses,
however SqliteHook's `get_conn` method is always calling the main class
attribute. Find the correct attribute and use this to establish the
connection.

Allow attr setting outside init for test case

Closes #10147

coopergillan committed:
Add type annotations for Sqlite (#10157)

Add type annotations for Sqlite's hook and operator.

Part of #9708

coopergillan committed:
Add type annotation to providers/jenkins (#9947)

Part of #9708

ephraimbuddy committed:
Add correct signatures for operators in google provider package (#10144)

potiuk committed:
Bring back code coverage (#10143)

Fixes #10138

potiuk committed:
Remove coverage HTML report artifacts (#10168)

potiuk committed:
Documentation artifact are also uploaded as GitHub Actions Artifacts (#10158)

vlasvlasvlas committed:
Add Apache Airflow CODE_OF_CONDUCT.md (#9715)

ypatankar committed:
Updated parameter definition docs for filesystem.py (#10159)

potiuk committed:
Disable wiki. (#10173)

mik-laj committed:
Remove stat_name_handler attribute in plugins.rst (#10174)

mik-laj committed:
Improve heading on Email Configuration page (#10175)

mik-laj committed:
Add docs for airflow config command (#10177)

* Add docs for airflow config command

* Update docs/howto/email-config.rst

ephraimbuddy committed:
Add correct signatures for operators in amazon provider package (#10167)

ephraimbuddy committed:
Enforce keyword only arguments on apache operators (#10170)

Commit:
prevent DAG callback exception from crashing scheduler (#10096)

turbaszek committed:
Improve handling Dataproc cluster creation with ERROR state (#9593)

Handle cluster in DELETING state

Extend tests

fixup! Extend tests

fixup! fixup! Extend tests

fixup! fixup! fixup! Extend tests

j-y-matsubara committed:
Fix BaseSensorOperator soft_fail mode to respect downstream tasks trigger_rule (#8867)

Fixes the BaseSensorOperator to make respect the trigger_rule in downstream tasks, when setting soft_fail="True".

LeonY1 committed:
Changes to all the constructors to remove the args argument (#10163)

kosteev committed:
Add thredup to list of Airflow users (#10198)

coopergillan committed:
Add type annotations to S3 hook module (#10164)

mik-laj committed:
Update guide for Google Cloud Secret Manager Backend (#10172)

OmairK committed:
Type annotation for Docker operator (#9733)

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

retornam committed:
Update JS packages to latest versions (#9811) (#9921)

Signed-off-by: Raymond Etornam <retornam@users.noreply.github.com>

ephraimbuddy committed:
Add correct signature to all operators and sensors (#10205)

* add correct signature to operators in providers package

* add keyword only to operators and sensors outside provider package

* remove unused type ignore

potiuk committed:
You can sync your fork master with apache/airflow master via UI (#10209)

We are using newly added feature of GitHub to add manually triggered
workflow to enable manually-triggered force-syncing of your fork
with apache/airflow.

potiuk committed:
Pylint checks should be way faster now (#10207)

* Pylint checks should be way faster now

Instead of running separate pylint checks for tests and main source
we are running a single check now. This is possible thanks to a
nice hack - we have pylint plugin that injects the right
"# pylint: disable=" comment for all test files while reading
the file content by astroid (just before tokenization)

Thanks to that we can also separate out pylint checks
to a separate job in CI - this way all pylint checks will
be run in parallel to all other checks effectively halfing
the time needed to get the static check feedback and potentially
cancelling other jobs much faster.

* fixup! Pylint checks should be way faster now

TreeKat71 committed:
Fix chart: parameterize namespace (#10213)

Replace fixed namespace "airflow" with variable {{ .Release.Namespace }}

mik-laj committed:
Docs: Separate page for each Secrets backend (#10211)

potiuk committed:
Improves stability of reported coverage and makes it nicer (#10208)

* Improves stability of reported coverage and makes it nicer

With this change we only upload coverage report in the case
when all tests were successuful and actually executed. This means
that coverage report will not be run when the job gets canceled.

Currently a lof of coverage reports gathered contain far less
coverage because when static check fails or docs some test jobs
could already submit their coverage - resulting in partial coverage
reports.

With this change we also remove comment from coverage report
and replace it with (for now) informational status message published
to github. If we see that it works, we can change it to a
PR-failing status if coverage drops for a given PR.

This way we might get our coverage monotonously increasing :).

rootcss committed:
Add airflow connections export command (#9856) (#10081)

msumit committed:
Handle IntegrityError while creating TIs (#10136)

While doing a trigger_dag from UI, DagRun gets created first and then WebServer starts creating TIs. Meanwhile, Scheduler also picks up the DagRun and starts creating the TIs, which results in IntegrityError as the Primary key constraint gets violated. This happens when a DAG has a good number of tasks.

Also, changing the TIs array with a set for faster lookups for Dags with too many tasks.

Swalloow committed:
Merge branch 'typing-aws-emr' of https://github.com/Swalloow/airflow into typing-aws-emr
